src/gurt/shm_alloc.c

```diff
@@ -0,0 +1,296 @@
+/**
+ * (C) Copyright 2024 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+
+#include <sys/syscall.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <pthread.h>
+
+#include <gurt/atomic.h>
+#include <gurt/common.h>
+#include <gurt/shm_alloc.h>
+#include <gurt/shm_utils.h>
+
+/* the name of shared memory used for mmap which will be found under /dev/shm/ */
+#define daos_shm_name "daos_shm_cache"
+
+/* pid of current process */
+static int          pid;
+
+/* the memory allocator that will be used to handle small memory allocation */
+static __thread int idx_small = -1;
+
+/* the address of shared memory region */
+struct d_shm_hdr   *d_shm_head;
+
+/* the attribute set for mutex located inside shared memory */
+pthread_mutexattr_t d_shm_mutex_attr;
+
+/* the number of times to try calling shm_open() */
+#define RETRY (5)
+
+/**
+ * pid of the process who creates shared memory region. shared memory is NOT unmapped when this
+ * process exits to keep shm always available. shared memory is unmapped when other processes
+ * exit.
+ */
+static int          pid_shm_creator;
+
+int
+shm_init(void)
+{
+	int      i;
+	int      shm_ht_fd;
+	int      shmopen_perm = 0600;
+	void    *shm_addr;
+	int      rc;
+	char     daos_shm_name_buf[64];
+	uint64_t shm_size;
+	uint64_t shm_pool_size;
+
+	if (pid == 0)
+		pid = getpid();
+	if (d_shm_head) {
+		while (d_shm_head->magic != DSM_MAGIC)
+			usleep(1);
+		/* shared memory already initlized in current process */
+		return 0;
+	}
+
+	rc = d_getenv_uint64_t("DAOS_SHM_SIZE", &shm_size);
+	if (rc != -DER_NONEXIST) {
+		/* set parameter from env */
+		shm_pool_size = shm_size / N_SHM_POOL;
+		if (shm_pool_size % 4096)
+			/* make shm_pool_size 4K aligned */
+			shm_pool_size += (4096 - (shm_pool_size % 4096));
+		shm_size = shm_pool_size * N_SHM_POOL + sizeof(struct d_shm_hdr);
```

Comment:
In fact my remarks on aligning the `struct d_shm_addr` was to have a sizeof struct which will be a multiple of the `PAGE_SIZE` otherwise `shm_size` could be not a multiple of `PAGE_SIZE` from my understanding. Then the shared pools (following the header), could be not aligned on PAGE_SIZE (from my understanding).
However, I have not check that the align attribute will properly change the sizeof. I will check this asap.

Reply:
"shm_size" does not have be a multiple of PAGE_SIZE. mmap() does not require size to be a multiple of PAGE_SIZE. The memory allocator will use some space in the pool too. I am not sure making shm_size a multiple of PAGE_SIZE will bring noticeable benefit. Maybe some performance tests could help to clarify later.

Reply:
From what I know it is indeed always better to have aligned memory for performance.
Moreover, from what I understand, the size really allocated by `mmap()` will be the same.
The only difference is that the padding will be added at the end of allocated memory instead of being between the `struct d_shm_hdr` and the first shm_pool.
In any case, this is not a blocker from my side.